### PR TITLE
[Refactoring] changing EventDeadLetter::store

### PR DESCRIPTION
**Pull request by nvduc91**
Changing EventDeadLetter::store signature from

```Mono<Void> store(Group registeredGroup, Event failDeliveredEvent, InsertionId insertionId);```
to
```Mono<InsertionId> store(Group registeredGroup, Event failDeliveredEvent);```
That way we ensure InsertionId unicity, disable (untested) upserts, and prevent bugs from happening.

### CODE REVIEW

**Comment by rouazana** [comment]
>  That way we ensure InsertionId unicity, disable (untested) upserts, and prevent bugs from happening.

This PR description is very interesting. It should be included in the commit comment.

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/EventDeadLettersRoutesTest.java [hunk 1]
```diff
@@ -905,10 +923,11 @@ void postRedeliverSingleEventShouldFailWhenInvalidAction() {
...
         @Test
         void postRedeliverSingleEventShouldFailWhenMissingAction() {
-            deadLetters.store(new EventBusTestFixture.GroupA(), EVENT_1, INSERTION_ID_1).block();
-
+            InsertionId insertionId = deadLetters.store(new EventBusTestFixture.GroupA(), EVENT_1).block();
+            assertThat(insertionId).isNotNull();
+            
             when()
-                .post("/events/deadLetter/groups/" + SERIALIZED_GROUP_A + "/" + INSERTION_UUID_1)
+                .post("/events/deadLetter/groups/" + SERIALIZED_GROUP_A + "/" + insertionId.getId().toString())
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/EventDeadLettersRoutesTest.java]: Don't we have a convenience `insertionId.asString()`?

**Comment by nvduc91** [comment]
That’s great.I should used that method.

Vào Th 5, 13 thg 2, 2020 lúc 16:11 Tellier Benoit <notifications@github.com>
đã viết:

File: mailbox/event/event-cassandra/src/main/java/org/apache/james/mailbox/events/CassandraEventDeadLetters.java [hunk 2]
```diff
@@ -39,13 +39,15 @@
...
     @Override
-    public Mono<Void> store(Group registeredGroup, Event failDeliveredEvent, InsertionId insertionId) {
+    public Mono<InsertionId> store(Group registeredGroup, Event failDeliveredEvent) {
         Preconditions.checkArgument(registeredGroup != null, REGISTERED_GROUP_CANNOT_BE_NULL);
         Preconditions.checkArgument(failDeliveredEvent != null, FAIL_DELIVERED_EVENT_CANNOT_BE_NULL);
-        Preconditions.checkArgument(insertionId != null, FAIL_DELIVERED_ID_INSERTION_CANNOT_BE_NULL);
 
-        return cassandraEventDeadLettersDAO.store(registeredGroup, failDeliveredEvent, insertionId)
+        InsertionId insertionId = InsertionId.random();
+        cassandraEventDeadLettersDAO.store(registeredGroup, failDeliveredEvent, insertionId)
```
Comment (mbaechler) [inline mailbox/event/event-cassandra/src/main/java/org/apache/james/mailbox/events/CassandraEventDeadLetters.java]: It's not working anymore: you should chain the return of `Mono.just(insertionId)` to the `store` Mono otherwise a `block` on the return of this method won't block on the actual saving into Cassandra

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/EventDeadLettersRoutesTest.java (hunk 1 above)
Comment (nvduc91) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/EventDeadLettersRoutesTest.java]: Actually we don't have this method but i will make it.

**Comment by nvduc91** [comment]
> Why do you think generating InsertionId internally ensure unicity?
> 
> BTW, the InsertionId as a parameter is not for upsert but of deduplication / idempotence ... but I can believe it's not tested
> 
> Finally, what are the bugs that it prevents?

For now, i can not answer your question but @chibenwa may do that.
The `description of PR` was suggested by him.

**Comment by chibenwa** [comment]
> Why do you think generating InsertionId internally ensure unicity?

As a caller I can no more specify the InsertionId myself.

The implementation can then ensure InsertionId unicity.

This "overwrite behavior is BTW not tested in previous code.

It seems legitimate to me to give DeadLetter the responsibility to generate the InsertionId.

> BTW, the InsertionId as a parameter is not for upsert but of deduplication / idempotence ... but I can believe it's not tested

IMO indempotence makes sense at the DAO level.

deduplication is not the concern here, more being sure that a reprocessed event will be well stored if the reprocessing fails. See https://issues.apache.org/jira/browse/MAILBOX-386

> Finally, what are the bugs that it prevents?

As a caller it prevents me from overwriting content in envent dead-letter.

It do not prevent bugs regarding current code, however it enforces correct API usage thus prevents future bugs from happening and thus is valuable.

File: mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDelivery.java [hunk 3]
```diff
@@ -121,7 +121,8 @@ private StoreToDeadLetters(Group group, EventDeadLetters eventDeadLetters) {
 
             @Override
             public Mono<Void> handle(Event event) {
-                return eventDeadLetters.store(group, event, EventDeadLetters.InsertionId.random());
+                eventDeadLetters.store(group, event);
+                return Mono.empty();
```
Comment (trantienduchn) [inline mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDelivery.java]:
```java
return eventDeadLetters.store(group, event)
    .then();
```

**Comment by Arsnael** [comment]
```
[d3fc11f8e91e5e626e8d02737dfc3b3edb4d462e] [ERROR] /james-project/mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/GroupConsumerRetry.java:[108,36] method store in interface org.apache.james.mailbox.events.EventDeadLetters cannot be applied to given types;
[d3fc11f8e91e5e626e8d02737dfc3b3edb4d462e]   required: org.apache.james.mailbox.events.Group,org.apache.james.mailbox.events.Event
[d3fc11f8e91e5e626e8d02737dfc3b3edb4d462e]   found: org.apache.james.mailbox.events.Group,org.apache.james.mailbox.events.Event,org.apache.james.mailbox.events.EventDeadLetters.InsertionId
[d3fc11f8e91e5e626e8d02737dfc3b3edb4d462e]   reason: actual and formal argument lists differ in length
[d3fc11f8e91e5e626e8d02737dfc3b3edb4d462e] [ERROR] /james-project/mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/GroupConsumerRetry.java:[127,49] method store in interface org.apache.james.mailbox.events.EventDeadLetters cannot be applied to given types;
[d3fc11f8e91e5e626e8d02737dfc3b3edb4d462e]   required: org.apache.james.mailbox.events.Group,org.apache.james.mailbox.events.Event
[d3fc11f8e91e5e626e8d02737dfc3b3edb4d462e]   found: org.apache.james.mailbox.events.Group,org.apache.james.mailbox.events.Event,org.apache.james.mailbox.events.EventDeadLetters.InsertionId
[d3fc11f8e91e5e626e8d02737dfc3b3edb4d462e]   reason: actual and formal argument lists differ in length
```

File: mailbox/event/event-cassandra/src/main/java/org/apache/james/mailbox/events/CassandraEventDeadLetters.java [hunk 4]
```diff
@@ -39,13 +39,14 @@
...
+    public Mono<InsertionId> store(Group registeredGroup, Event failDeliveredEvent) {
         Preconditions.checkArgument(registeredGroup != null, REGISTERED_GROUP_CANNOT_BE_NULL);
         Preconditions.checkArgument(failDeliveredEvent != null, FAIL_DELIVERED_EVENT_CANNOT_BE_NULL);
-        Preconditions.checkArgument(insertionId != null, FAIL_DELIVERED_ID_INSERTION_CANNOT_BE_NULL);
 
+        InsertionId insertionId = InsertionId.random();
         return cassandraEventDeadLettersDAO.store(registeredGroup, failDeliveredEvent, insertionId)
-            .then(cassandraEventDeadLettersGroupDAO.storeGroup(registeredGroup));
+            .then(cassandraEventDeadLettersGroupDAO.storeGroup(registeredGroup))
+            .then(Mono.just(insertionId));
```
Comment (trantienduchn) [inline mailbox/event/event-cassandra/src/main/java/org/apache/james/mailbox/events/CassandraEventDeadLetters.java]: `thenReturn(insertionId)` :)

File: mailbox/event/event-cassandra/src/main/java/org/apache/james/mailbox/events/CassandraEventDeadLettersDAO.java [hunk 5]
```diff
@@ -85,11 +85,12 @@ private PreparedStatement prepareSelectInsertionIdsWithGroupStatement(Session se
             .where(eq(GROUP, bindMarker(GROUP))));
     }
 
-    Mono<Void> store(Group group, Event failedEvent, EventDeadLetters.InsertionId insertionId) {
-        return executor.executeVoid(insertStatement.bind()
+    Mono<EventDeadLetters.InsertionId> store(Group group, Event failedEvent, EventDeadLetters.InsertionId insertionId) {
```
Comment (trantienduchn) [inline mailbox/event/event-cassandra/src/main/java/org/apache/james/mailbox/events/CassandraEventDeadLettersDAO.java]: is this change necessary? I don't see the any `.map()` after `.store()`

**Comment by rouazana** [comment]
> This PR description is very interesting. It should be included in the commit comment.

Can you modify your commit please?

**Comment by nvduc91** [comment]
> > This PR description is very interesting. It should be included in the commit comment.
> 
> Can you modify your commit please?

How i should changing it?

**Comment by rouazana** [comment]
> How i should changing it?

You do a git rebase, and there you can modify the comment title.
In this specific case as you have only one commit you can also simple do a git commit --amend

**Comment by nvduc91** [comment]
> In this specific case as you have only one commit you can also simple do a git commit --amend

Ah, i got it.

**Comment by mbaechler** [comment]
> > Why do you think generating InsertionId internally ensure unicity?
> 
> As a caller I can no more specify the InsertionId myself.
> 
> The implementation can then ensure InsertionId unicity.

Ok, you moved the responsibility. But the system is as before: we generate an id. We can mess with that id in both cases.
 
> This "overwrite behavior is BTW not tested in previous code.
> 
> It seems legitimate to me to give DeadLetter the responsibility to generate the InsertionId.
> 
> > BTW, the InsertionId as a parameter is not for upsert but of deduplication / idempotence ... but I can believe it's not tested
> 
> IMO indempotence makes sense at the DAO level.

Why?

As a REST API user, idempotence makes sense too. There are many things you can't do if id is not generated soon enough.

> deduplication is not the concern here, more being sure that a reprocessed event will be well stored if the reprocessing fails. See https://issues.apache.org/jira/browse/MAILBOX-386

You describe the initial concern but it doesn't change the fact that having an external InsertionId is valuable to prevent duplication.

> > Finally, what are the bugs that it prevents?
> 
> As a caller it prevents me from overwriting content in envent dead-letter.

I agree that preventing overwriting an event would require a LWT with Cassandra and it would make everything slower.

> It do not prevent bugs regarding current code, however it enforces correct API usage thus prevents future bugs from happening and thus is valuable.

As it's LWT vs internal insertion id generation, I agree to choose the second one.

**Comment by trantienduchn** [comment]
```
[688c663009f9052f0a91028a8848cc9cffa63117] [ERROR] Failures: 
[688c663009f9052f0a91028a8848cc9cffa63117] [ERROR]   RabbitMQWebAdminServerTaskSerializationIntegrationTest.postRedeliverSingleEventShouldCreateATask:534 1 expectation failed.
[688c663009f9052f0a91028a8848cc9cffa63117] JSON path status doesn't match.
[688c663009f9052f0a91028a8848cc9cffa63117] Expected: is "failed"
[688c663009f9052f0a91028a8848cc9cffa63117]   Actual: completed
```

this is an unstable test, not from your changes @nvduc91

**Comment by nvduc91** [comment]
test this please

**Comment by trantienduchn** [comment]
```
[688c663009f9052f0a91028a8848cc9cffa63117] [ERROR] Failures: 
[688c663009f9052f0a91028a8848cc9cffa63117] [ERROR]   RabbitMQWebAdminServerTaskSerializationIntegrationTest.postRedeliverSingleEventShouldCreateATask:534 1 expectation failed.
[688c663009f9052f0a91028a8848cc9cffa63117] JSON path status doesn't match.
[688c663009f9052f0a91028a8848cc9cffa63117] Expected: is "failed"
[688c663009f9052f0a91028a8848cc9cffa63117]   Actual: completed
```

Again, I might be wrong, sorry, can you run it on your local? @nvduc91

**Comment by chibenwa** [comment]
test this please

**Comment by nvduc91** [comment]
> ```
> [688c663009f9052f0a91028a8848cc9cffa63117] [ERROR] Failures: 
> [688c663009f9052f0a91028a8848cc9cffa63117] [ERROR]   RabbitMQWebAdminServerTaskSerializationIntegrationTest.postRedeliverSingleEventShouldCreateATask:534 1 expectation failed.
> [688c663009f9052f0a91028a8848cc9cffa63117] JSON path status doesn't match.
> [688c663009f9052f0a91028a8848cc9cffa63117] Expected: is "failed"
> [688c663009f9052f0a91028a8848cc9cffa63117]   Actual: completed
> ```
> 
> Again, I might be wrong, sorry, can you run it on your local? @nvduc91

I will check it.

**Comment by nvduc91** [comment]
test this please

**Comment by nvduc91** [comment]
test this please

**Comment by trantienduchn** [comment]
```
[688c663009f9052f0a91028a8848cc9cffa63117] [ERROR] Errors: 
[688c663009f9052f0a91028a8848cc9cffa63117] [ERROR]   MemorySpamAssassinContractTest.spamShouldBeDeliveredInSpamMailboxOrInboxWhenMultipleRecipientsConfigurations » ConditionTimeout
```

test this please

**Comment by Arsnael** [comment]
```
[688c663009f9052f0a91028a8848cc9cffa63117] [ERROR] Failures: 
[688c663009f9052f0a91028a8848cc9cffa63117] [ERROR]   RabbitMQWebAdminServerTaskSerializationIntegrationTest.postRedeliverSingleEventShouldCreateATask:534 1 expectation failed.
[688c663009f9052f0a91028a8848cc9cffa63117] JSON path status doesn't match.
[688c663009f9052f0a91028a8848cc9cffa63117] Expected: is "failed"
[688c663009f9052f0a91028a8848cc9cffa63117]   Actual: completed
[688c663009f9052f0a91028a8848cc9cffa63117] 
[688c663009f9052f0a91028a8848cc9cffa63117] [INFO] 
[688c663009f9052f0a91028a8848cc9cffa63117] [ERROR] Tests run: 172, Failures: 1, Errors: 0, Skipped: 2
```

File: server/protocols/webadmin-integration-test/distributed-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/rabbitmq/RabbitMQWebAdminServerTaskSerializationIntegrationTest.java [hunk 6]
```diff
@@ -511,15 +512,17 @@ void postRedeliverSingleEventShouldCreateATask(GuiceJamesServer guiceJamesServer
...
             .build();
 
-        guiceJamesServer
+        EventDeadLetters.InsertionId insertionId = guiceJamesServer
             .getProbe(EventDeadLettersProbe.class)
             .getEventDeadLetters()
             .store(group, event)
             .block();
 
+        Assertions.assertThat(insertionId).isNotNull();
```
Comment (chibenwa) [inline server/protocols/webadmin-integration-test/distributed-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/rabbitmq/RabbitMQWebAdminServerTaskSerializationIntegrationTest.java]: Please remove this assertion: it's not the main assertion, it makes the test case harder to read.

File: server/protocols/webadmin-integration-test/distributed-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/rabbitmq/RabbitMQWebAdminServerTaskSerializationIntegrationTest.java [hunk 7]
```diff
@@ -749,4 +738,15 @@ void cassandraMappingsSolveInconsistenciesShouldComplete() {
             .body("additionalInformation.successfulMappingsCount", is(0))
             .body("additionalInformation.errorMappingsCount", is(0));
     }
+
+    private MailboxListener.MailboxAdded createMailboxAdded() {
+        final String uuid = "6e0dd59d-660e-4d9b-b22f-0354479f47b4";
```
Comment (chibenwa) [inline server/protocols/webadmin-integration-test/distributed-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/rabbitmq/RabbitMQWebAdminServerTaskSerializationIntegrationTest.java]: We avoid using final variables

Comment (mbaechler) [reply server/protocols/webadmin-integration-test/distributed-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/rabbitmq/RabbitMQWebAdminServerTaskSerializationIntegrationTest.java]: because we don't use variable reuse but in very rare cases so it protects us from problems we should not have to deal with while adding noise.

**Comment by Arsnael** [comment]
Merged
